File: ComposedSet/ComposedString.fs
﻿namespace ComposedSet.FSharp

open ComposedSet
open System.Text.RegularExpressions

type StringComposedSetDatabase() =
    inherit ComposedSetDatabase<string>()

    override this.Compose indicies = 
        indicies |> Seq.map (fun i -> this.parts.[i]) |> String.concat ""
            
    override this.Split composed = 
         Regex.Split(composed, @"(\.)|(\/)|(\@)|(_)|(\f)", RegexOptions.Compiled) |> Array.filter (fun s -> not (System.String.IsNullOrEmpty s))

File: ComposedSet_CSharp/ComposedString.cs
using System.Collections.Generic;
using System.Text;
using System.Linq;
using System.Text.RegularExpressions;

namespace ComposedSet.CSharp
{
    using ComposedString = ComposedSet<string, StringComposedSetDatabase>;
    public class StringComposedSetDatabase : BaseComposedSetDatabase<string>
    {
        const string regexSplitter = @"(\.)|(\/)|(\@)|(_)|(\f)";
        static Regex regex = new Regex(regexSplitter, RegexOptions.Compiled);
        public override string[] Split(string composed)
        {
            return regex.Split(composed).Where(s => !string.IsNullOrEmpty(s)).ToArray();
        }

        public override string Compose(List<int> indices)
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0, length = indices.Count; i < length; ++i)
            {
                sb.Append(Parts[indices[i]]);
            }
            return sb.ToString();
        }
    }
}